Library/Homebrew/test/api/internal_tap_json/formula_spec.rb

```diff
@@ -0,0 +1,128 @@
+    end
+
+    it "creates the expected hash" do
+      api_hash = CoreTap.instance.to_api_hash
+      api_hash["tap_git_head"] = tap_git_head # tricky to mock
+
+      expect(JSON.generate(api_hash)).to eq(internal_tap_json)
+    end
+  end
+
+  context "when loading JSON" do
+    before do
+      ENV["HOMEBREW_INTERNAL_JSON_V3"] = "1"
+      ENV.delete("HOMEBREW_NO_INSTALL_FROM_API")
+
+      allow(Homebrew::API).to receive(:fetch_json_api_file)
+        .with("internal/v3/homebrew-core.jws.json")
+        .and_return(JSON.parse(internal_tap_json), false)
+    end
+
+    it "loads tap aliases" do
+      expect(CoreTap.instance.alias_table).to eq({
+        "fennel-lang" => "fennel",
+        "ponyc-lang"  => "ponyc",
+      })
+    end
+
+    it "loads formula renames" do
+      expect(CoreTap.instance.formula_renames).to eq({
+        "advancemenu" => "advancemame",
+        "amtk"        => "libgedit-amtk",
+        "annie"       => "lux",
+        "antlr2"      => "antlr@2",
+        "romanesco"   => "fennel",
+      })
+    end
+
+    it "loads tap migrations" do
+      expect(CoreTap.instance.tap_migrations).to eq({
+        "adobe-air-sdk"          => "homebrew/cask",
+        "android-ndk"            => "homebrew/cask",
+        "android-platform-tools" => "homebrew/cask",
+        "android-sdk"            => "homebrew/cask",
+        "app-engine-go-32"       => "homebrew/cask/google-cloud-sdk",
+      })
+    end
+
+    it "loads tap git head" do
+      expect(Homebrew::API::Formula.tap_git_head)
+        .to eq(tap_git_head)
+    end
+
+    context "when loading formulae" do
+      let(:fennel_metadata) do
+        {
+          "dependencies"     => ["lua"],
+          "desc"             => "Lua Lisp Language",
+          "full_name"        => "fennel",
+          "homepage"         => "https://fennel-lang.org",
+          "installed"        => [],
+          "license"          => "MIT",
+          "name"             => "fennel",
+          "ruby_source_path" => "Formula/f/fennel.rb",
+          "tap"              => "homebrew/core",
+          "tap_git_head"     => tap_git_head,
+          "versions"         => { "bottle"=>true, "head"=>nil, "stable"=>"1.4.0" },
+        }
+      end
+
+      let(:ponyc_metadata) do
+        {
+          "desc"                   => "Object-oriented, actor-model, capabilities-secure programming language",
+          "full_name"              => "ponyc",
+          "homepage"               => "https://www.ponylang.io/",
+          "license"                => "BSD-2-Clause",
+          "name"                   => "ponyc",
+          "ruby_source_path"       => "Formula/p/ponyc.rb",
+          "tap"                    => "homebrew/core",
+          "tap_git_head"           => tap_git_head,
+          "uses_from_macos"        => [{ "llvm"=>[:build, :test] }, "zlib"],
+          "uses_from_macos_bounds" => [{}, {}],
+          "versions"               => { "bottle"=>true, "head"=>nil, "stable"=>"0.58.1" },
```

Comment:
```suggestion
          "versions"               => { "bottle"=>true, "stable"=>"0.58.1" },
```

Reply:
Yeah, I can probably compact or avoid adding that field in the first place.

Reply:
Never mind, this only exists in the public facing JSON and should already be omitted from the internal one. I'm just using the public JSON here for testing because it exercises more parts of the formula code.

Reply:
Yeh, I think it'd be preferable to use the internal JSON and to get full coverage of the formula code through the input JSON.

Reply:
> Yeh, I think it'd be preferable to use the internal JSON and to get full coverage of the formula code through the input JSON.

Sorry, I don't know what you mean by this. How can we get full coverage of the formula code through the input JSON? What input JSON are we referring to here?

Reply:
I'm saying it seems very weird to be to use the non-internal public JSON for testing the internal JSON format?

Reply:
It's not that weird. It essentially mimics `brew readall`.

Reply:
@apainintheneck Ok but aren't these JSON formats going to diverge significantly? 

I think I might just need an ELI5 treatment here, sorry.

Reply:
This is basically a fancy `to_hash` test. We could do it as a series of checks instead, though would be multiple `expects` in a single check so might be long:

```rb
expect(formula.name).to eq("something")
expect(formula.version.to eq("1.0")
expect(formula.revision to be_zero)
# ...
```

Reply:
`#to_hash` is a bit better than doing a bunch of individual checks since it will end up calling more methods internally to fill out the fields we're not checking.

The goal here is not to validate the `#to_hash` method but instead to be reasonably confident that the formula loads without failing. `#to_hash` is just an easy way to take a peek inside the formula instance and run some arbitrary methods along the way. I'm open to alternatives that would be clearer, briefer or provide more assurance of correctness. The current test is not comprehensive at all.

`#to_api_hash` is already being exercised by a different test where we generate the JSON API for the core formula tap.